src/SignalR/common/Http.Connections/src/Features/IBeforeShutdown.cs

```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Http.Connections.Features
+{
+    /// <summary>
+    /// Feature that allows registering functions to run on application closing before connections are closed.
+    /// </summary>
+    public interface IBeforeShutdown
+    {
+        /// <summary>
+        /// Register a function to run when graceful shutdown starts and before connections are closed.
+        /// </summary>
+        /// <remarks>
+        /// The Server or Host may forcefully close connections before this task completes.
+        /// </remarks>
+        /// <param name="callback">Function to run on graceful shutdown.</param>
+        /// <returns>Disposable that will remove the registered function.</returns>
+        IDisposable Register(Func<Task> callback);
+
+        /// <summary>
+        /// Trigger the registered callbacks.
+        /// </summary>
+        /// <returns>Task representing the callbacks running.</returns>
+        Task TriggerAsync();
```

Comment:
Does this need to be part of the public API? I was thinking that the IBeforeShutdown implementation cool register the callbacks with an internal service type that we use to trigger the callbacks.

Reply:
If it isn't part of the API anyone can come along and replace the implementation and now we have no way of running the callbacks

Reply:
That could be said for any service with internal dependencies. Anyone who wants the default behavior should be registering with the default service. They can register another service that calls into this if they want.

If the way this API was designed to be used was by replacing the service, we could have *just* `TriggerAsync()` on the interface and get rid of `Register(Func<Task>)`

Reply:
Why did we decide against something like `IDisposable ConnectionOptions.RegisterBeforeShutdown(Func<Task>)` again?

Reply:
I remember. We wanted this to be usable inside of Hub.OnConnectedAsync() and mutating options felt weird there. I still think leaving TriggerAsync() on the public service type is weird though.

Reply:
> I still think leaving TriggerAsync() on the public service type is weird though.

Agreed, we can decide the interface isn't meant for replacing and remove it